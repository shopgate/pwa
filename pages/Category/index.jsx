import React, { PureComponent } from 'react';
import Consume from '@shopgate/pwa-common/components/Consume';
import { hex2bin } from '@shopgate/pwa-common/helpers/data';
import { RouteContext } from '@shopgate/pwa-common/context';
import { View } from '@shopgate/engage/components';
import { themeConfig } from '@shopgate/pwa-common/helpers/config';
import CategoryContent from './components/Content';

const { colors } = themeConfig;

const map = {
  id: 'params.categoryId',
  open: 'open',
};

/**
 * The category page component.
 */
class Category extends PureComponent {
  /**
   * @param {Object} props the consumed props.
   * @returns {JSX}
   */
  consumeRenderer = ({ id, open }) => {
    if (!open) {
      return null;
    }

    return <CategoryContent categoryId={hex2bin(id) || null} />;
  }

  /**
   * @returns {JSX}
   */
  render() {
    return (
      <View background={colors.background} aria-hidden={false}>
        <Consume context={RouteContext} props={map}>
          {this.consumeRenderer}
        </Consume>
      </View>
    );
  }
}

export default Category;
