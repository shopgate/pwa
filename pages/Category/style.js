import { css } from 'glamor';

const headlineWrapper = css({
  marginTop: 12,
}).toString();

export default {
  headlineWrapper,
};
