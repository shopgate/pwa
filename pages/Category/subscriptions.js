import { ACTION_PUSH, ACTION_REPLACE } from '@virtuous/conductor';
import { buildFetchCategoryProductsParams, getProductsResult } from '@shopgate/engage/product';
import { getCurrentRoute } from '@shopgate/pwa-common/selectors/router';
import { historyPop } from '@shopgate/pwa-common/actions/router';
import expireProductsByHash from '@shopgate/pwa-common-commerce/product/action-creators/expireProductsByHash';
import fetchCategory from '@shopgate/pwa-common-commerce/category/actions/fetchCategory';
import fetchCategoryProducts from '@shopgate/pwa-common-commerce/category/actions/fetchCategoryProducts';
import fetchFilters from '@shopgate/pwa-common-commerce/filter/actions/fetchFilters';
import { hex2bin } from '@shopgate/pwa-common/helpers/data';
import {
  categoryWillEnter$,
  categoryDidEnter$,
  categoryProductsNeedUpdate$,
  errorVisibleCategory$,
} from './streams';

/**
 * Category subscriptions.
 * @param {Function} subscribe The subscribe function.
 */
export default function category(subscribe) {
  subscribe(categoryWillEnter$, ({ dispatch, action, getState }) => {
    if (![ACTION_PUSH, ACTION_REPLACE].includes(action?.historyAction)) {
      return;
    }

    const { filters, offset = 0 } = action.route.state;
    const categoryId = hex2bin(action.route.params.categoryId);

    const { hash, expired } = getProductsResult(getState(), { categoryId });

    if (expired) {
      dispatch(expireProductsByHash(hash));
    }

    dispatch(fetchCategory(categoryId));
    dispatch(fetchCategoryProducts({
      categoryId,
      filters,
      offset,
      ...buildFetchCategoryProductsParams(),
    }));
  });

  subscribe(categoryDidEnter$, ({ dispatch }) => {
    dispatch(fetchFilters());
  });

  /**
   * Gets triggered on pipeline category error.
   */
  subscribe(errorVisibleCategory$, ({ dispatch }) => {
    dispatch(historyPop());
  });

  subscribe(categoryProductsNeedUpdate$, ({ action, dispatch, getState }) => {
    const { params, state: { offset = 0 } } = getCurrentRoute(getState());
    const categoryId = hex2bin(params.categoryId);
    const { filters } = action;

    dispatch(fetchCategoryProducts({
      categoryId,
      filters,
      offset,
      ...buildFetchCategoryProductsParams(),
    }));
  });
}
