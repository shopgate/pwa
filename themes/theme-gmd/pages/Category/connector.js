import { connect } from 'react-redux';
import {
  getCategoryProductCount,
  getCurrentCategory,
  getCurrentCategoryId,
} from '@shopgate/pwa-common-commerce/category/selectors';
import { getProductsResult } from '@shopgate/pwa-common-commerce/product/selectors/product';
import { isFilterBarShown } from './selectors';

/**
 * Maps the contents of the state to the component props.
 * @param {Object} state The current application state.
 * @param {Object} props The component props.
 * @return {Object} The extended component props.
 */
const mapStateToProps = (state, props) => ({
  category: getCurrentCategory(state, props),
  hasProducts: getCategoryProductCount(state, props) > 0,
  isFilterBarShown: isFilterBarShown(state, props),
  isRoot: !getCurrentCategoryId(state, props),
  ...getProductsResult(state, props),
});

export default connect(mapStateToProps, null, null, { withRef: true });
