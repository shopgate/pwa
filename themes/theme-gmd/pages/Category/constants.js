export const GRID_VIEW = 'grid';
export const LIST_VIEW = 'list';

export const SET_CATEGORY_VIEW_MODE = 'SET_CATEGORY_VIEW_MODE';
