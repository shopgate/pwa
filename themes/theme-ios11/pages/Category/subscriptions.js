import getCategory from '@shopgate/pwa-common-commerce/category/actions/getCategory';
import { getCurrentCategoryId } from '@shopgate/pwa-common-commerce/category/selectors';
import {
  categoryRouteDidEnter$,
  categoryError$,
} from '@shopgate/pwa-common-commerce/category/streams';
import showModal from '@shopgate/pwa-common/actions/modal/showModal';
import goBackHistory from '@shopgate/pwa-common/actions/history/goBackHistory';

/**
 * Filter subscriptions.
 * @param {Function} subscribe The subscribe function.
 */
export default function category(subscribe) {
  /**
   * Gets triggered on entering the filter route.
   */
  subscribe(categoryRouteDidEnter$, ({ dispatch, getState }) => {
    const state = getState();
    dispatch(getCategory(getCurrentCategoryId(state)));
  });

  /**
   * Gets triggered on pipeline category error.
   */
  subscribe(categoryError$, ({ action, dispatch }) => {
    const { errorCode } = action;
    let message = 'modal.body_error';

    if (errorCode === 'ENOTFOUND') {
      message = 'category.error.not_found';
    }

    dispatch(showModal({
      confirm: 'modal.ok',
      dismiss: null,
      message,
      title: 'category.error.title',
    }));
    dispatch(goBackHistory(1));
  });
}
