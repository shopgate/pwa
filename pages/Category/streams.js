import { main$ } from '@shopgate/pwa-common/streams/main';
import { routeWillEnter$, routeDidEnter$, routeWillLeave$ } from '@shopgate/pwa-common/streams/router';
import { CATEGORY_PATH, RECEIVE_CATEGORY, ERROR_CATEGORY } from '@shopgate/pwa-common-commerce/category/constants';
import { filtersDidUpdate$ } from '@shopgate/pwa-common-commerce/filter/streams';
import { getCurrentRoute } from '@shopgate/pwa-common/selectors/router';
import { hex2bin } from '@shopgate/pwa-common/helpers/data';

export const categoryWillEnter$ = routeWillEnter$
  .filter(({ action }) => action.route.pattern === `${CATEGORY_PATH}/:categoryId`);

export const categoryDidEnter$ = routeDidEnter$
  .filter(({ action }) => action.route.pattern === `${CATEGORY_PATH}/:categoryId`);

export const categoryWillLeave$ = routeWillLeave$
  .filter(({ action }) => action.route.pattern === `${CATEGORY_PATH}/:categoryId`);

/**
 * Filters stream state by type.
 * @param {Object} data The stream data.
 * @param {string} type The action type to filter.
 * @returns {boolean}
 */
const filterDataByType = (data, type) => {
  const { action, getState } = data;

  if (action.type !== type) {
    return false;
  }

  if (typeof action.categoryId === 'undefined') {
    return false;
  }

  const route = getCurrentRoute(getState());

  if (!route.params.categoryId) {
    return false;
  }

  return action.categoryId === hex2bin(route.params.categoryId);
};

export const receivedVisibleCategory$ = main$
  .filter(data => filterDataByType(data, RECEIVE_CATEGORY));

export const errorVisibleCategory$ = main$
  .filter(data => filterDataByType(data, ERROR_CATEGORY));

export const categoryFiltersDidUpdate$ = filtersDidUpdate$
  .filter(({ getState }) => {
    const { pattern } = getCurrentRoute(getState());
    return (pattern === `${CATEGORY_PATH}/:categoryId`);
  });
