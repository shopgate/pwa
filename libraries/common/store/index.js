import { composeWithDevTools } from 'redux-devtools-extension';
import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import { persistState } from '@virtuous/redux-persister';
import benchmarkMiddleware from '@shopgate/pwa-benchmark/profilers/redux';
import benchmarkController from '@shopgate/pwa-benchmark';
import persistedReducers from '../collections/PersistedReducers';
import initSubscribers from '../subscriptions';
import appConfig from '../helpers/config';
import streams from './middelwares/streams';
import logger from './middelwares/logger';

/**
 * The current version of the state created by this reducer.
 * @type {string}
 */
const STATE_VERSION = 'v1';
const storeKey = `shopgate-connect_${STATE_VERSION}`;

/**
 * Returns a normalised initialState from the localstorage.
 * @returns {Object}
 */
function getInitialState() {
  if (!window.localStorage) {
    return undefined;
  }

  const storedState = window.localStorage.getItem(storeKey);

  if (!storedState) {
    return undefined;
  }

  const normalisedState = storedState.replace(new RegExp('"isFetching":true', 'g'), '"isFetching":false');
  return JSON.parse(normalisedState);
}

/**
 * Configures the redux store with all it's middleware and enhancers.
 * @param {Function} reducers The reducers from the theme.
 * @param {Array} subscribers The subscribers to the streams middleware.
 * @return {Object} The redux store.
 */
export function configureStore(reducers, subscribers) {
  // Starts benchmark controller BEFORE adding the middleware.
  if (appConfig.benchmark) {
    benchmarkController.startup();
  }

  const store = createStore(
    reducers,
    getInitialState(),
    composeWithDevTools(
      applyMiddleware(...[
        thunk,
        ...appConfig.benchmark ? [benchmarkMiddleware] : [],
        streams,
        logger,
      ]),
      persistState({
        key: storeKey,
        paths: persistedReducers.getAll(),
      })
    )
  );

  initSubscribers(subscribers);

  if (window.Cypress) {
    window.store = store;
  }

  return store;
}

/**
 * Configures a mocked store for tests.
 * @param {Function} reducers The reducers to use in the mock store.
 * @param {Array} subscribers Streams subscribers to initialize.
 * @returns {Store}
 */
export function createMockStore(reducers = () => {}, subscribers = []) {
  const store = createStore(
    reducers,
    undefined,
    applyMiddleware(thunk, streams)
  );

  if (subscribers.length !== 0) {
    initSubscribers([].concat(subscribers));
  }

  return store;
}
