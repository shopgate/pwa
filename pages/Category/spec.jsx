import React from 'react';
import { shallow } from 'enzyme';
import { Provider } from 'react-redux';
import { createMockStore } from '@shopgate/pwa-common/store';
import Category from './index';

const store = createMockStore();
jest.mock('../../components/View/context.js');

describe('Pages: <Category />', () => {
  it('should render', () => {
    const wrapper = shallow((
      <Provider store={store}>
        <Category />
      </Provider>
    ));
    expect(wrapper).toMatchSnapshot();
  });
});
