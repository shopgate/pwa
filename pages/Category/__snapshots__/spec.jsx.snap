// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pages: <Category /> should render 1`] = `<Category />`;
